pilot/pkg/model/validation.go

```diff
@@ -1448,6 +1448,20 @@ func ValidateServiceRole(name, namespace string, msg proto.Message) error {
 		if len(rule.Services) == 0 {
 			errs = appendErrors(errs, fmt.Errorf("at least 1 service must be specified for rule %d", i))
 		}
+		// Regular rules and not rules (e.g. methods and not_methods should not be defined together).
+		sameAttributeKindError := "cannot have both regular and *not* attributes for the same kind (%s) for rule %d"
+		if len(rule.Hosts) > 0 && len(rule.NotHosts) > 0 {
+			errs = appendErrors(errs, fmt.Errorf(sameAttributeKindError, "i.e. hosts and not_hosts", i))
+		}
+		if len(rule.Methods) > 0 && len(rule.NotMethods) > 0 {
+			errs = appendErrors(errs, fmt.Errorf(sameAttributeKindError, "i.e. methods and not_methods", i))
+		}
+		if len(rule.Paths) > 0 && len(rule.NotPaths) > 0 {
+			errs = appendErrors(errs, fmt.Errorf(sameAttributeKindError, "i.e. paths and not_paths", i))
+		}
+		if len(rule.Ports) > 0 && len(rule.NotPorts) > 0 {
```

Comment:
You can also check if the port is valid, like if it can be converted to int and in the range `[0, 65535]`

Reply:
In the function call chain for port, eventually convertToPort() in rbac.go will be called, which does the checking. I'd say change the type in the API, from int32 to uint32 so Go compiler can catch it sooner.

Reply:
It's different. rbac.go checks it in runtime and only output the error message in logs and doesn't stop it being applied., it's hard for user to see it. here it checks in validation time which gives the user a message right after the command line and stop it from being applied. Whenever possible, we should try to do this in validation time.

Reply:
Ah, I see. But why not just change the date type to uint32 instead of int32? That way we don't have to add any logic code to validation. Wdyt?

Reply:
You will still need to check it's in the range. I'm okay to change it to an integer but I think we could just keep the API as-is just in case if we in the future want to support named ports like authn (https://preliminary.istio.io/docs/reference/config/istio.authentication.v1alpha1/#PortSelector)

Reply:
I see. Thanks. Will add a function to check